vertical-pod-autoscaler/pkg/utils/status/status_object.go

```diff
@@ -0,0 +1,163 @@
+	typedcoordinationv1 "k8s.io/client-go/kubernetes/typed/coordination/v1"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	// Parameters for retrying with exponential backoff.
+	retryBackoffInitialDuration = 100 * time.Millisecond
+	retryBackoffFactor          = 3
+	retryBackoffJitter          = 0
+	retryBackoffSteps           = 6
+)
+
+// Client for the status object.
+type Client struct {
+	client               typedcoordinationv1.LeaseInterface
+	name                 string
+	namespace            string
+	leaseDurationSeconds int32
+}
+
+// NewClient returns a client for the status object.
+func NewClient(c clientset.Interface, name, namespace string, leaseDuration time.Duration) *Client {
+	return &Client{
+		client:               c.CoordinationV1().Leases(namespace),
+		name:                 name,
+		namespace:            namespace,
+		leaseDurationSeconds: int32(leaseDuration) / int32(time.Second),
+	}
+}
+
+// UpdateStatus renews status object lease.
+// Status object will be created if it doesn't exist.
+func (c *Client) UpdateStatus() error {
+	updateFn := func() error {
+		lease, err := c.client.Get(c.name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			// Create lease if it doesn't exist.
+			return c.create()
+		} else if err != nil {
+			return err
+		}
+		lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+		_, err = c.client.Update(lease)
+		if apierrors.IsConflict(err) {
+			// Lease was updated by an another replica of the component.
+			// No error should be returned.
+			return nil
+		}
+		return err
+	}
+	return retryWithExponentialBackOff(updateFn)
+}
+
+func (c *Client) create() error {
+	_, err := c.client.Create(c.newLease())
+	if apierrors.IsAlreadyExists(err) {
+		// Lease was created by an another replica of the component.
+		// No error should be returned.
+		return nil
+	}
+	return err
+}
+
+func (c *Client) newLease() *apicoordinationv1.Lease {
+	return &apicoordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.name,
+			Namespace: c.namespace,
+		},
+		Spec: apicoordinationv1.LeaseSpec{
+			HolderIdentity:       pointer.StringPtr(c.name),
+			LeaseDurationSeconds: pointer.Int32Ptr(c.leaseDurationSeconds),
+		},
+	}
+}
+
+// GetStatus returns status object.
+func (c *Client) GetStatus() (*apicoordinationv1.Lease, error) {
+	var lease *apicoordinationv1.Lease
+	getFn := func() error {
+		var err error
+		lease, err = c.client.Get(c.name, metav1.GetOptions{})
+		return err
+	}
+	err := retryWithExponentialBackOff(getFn)
+	return lease, err
+}
+
+// IsStatusValid verifies if status was updated before lease timing out.
+func IsStatusValid(status *apicoordinationv1.Lease, leaseTimeout time.Duration) bool {
+	return status.CreationTimestamp.Add(leaseTimeout).After(time.Now()) ||
+		status.Spec.RenewTime.Add(leaseTimeout).After(time.Now())
+}
+
+func isRetryableAPIError(err error) bool {
+	// These errors may indicate a transient error that we can retry in tests.
```

Comment:
Why does it say we can retry in tests?

Reply:
Copy-paste error. I took list of retryable error from Cluster Loader code.